### PR TITLE
Normative: Adding Intl.ListFormat specification.

### LINKED ISSUE #33: List formatting API
A common scenario is to have a list of items (Array in JS) which has to be formatted in a given language. CLDR provides all the patterns for that.

Example use cases:
- List of participants in a chat
- List of days for which an alarm is set
- List of affected events in Calendar app

Proposed API:

``` javascript
var f = new Intl.ListFormat(navigator.languages, {
  'type': 'regular', // regular or duration
  'variant': 'regular', // regular, short, narrow
});

f.format(['John', 'Amy', 'Nick']); // "John, Amy and Nick" in en-US
```

this would also supersede current `Array.prototype.toLocaleString` with ability to build better localized lists.

I can imagine this not having it's own object and only extending Array.prototype.toLocaleString with options.

### PR DESCRIPTION
**Pull request by zbraniecki**
Fixes #33.

Based on https://github.com/tc39/proposal-intl-list-format which is ready for Stage 4.

### CODE REVIEW

**Comment by sffc** [comment]
Can you add some housekeeping changes with this PR:

- Add Intl.ListFormat to [Appendix A](https://tc39.es/ecma402/#annex-implementation-dependent-behaviour)
- Add to [Section 4](https://tc39.es/ecma402/#sec-api-overview)

With the above, if any of your other recent proposals are also missing, go ahead and add them.  I think it's fine if you do that in this PR, but you could make another PR if you or @gibson042 prefers.

**Comment by zbraniecki** [comment]
Added. IIRC we didn't add `Array.toLocaleString` to use `ListFormat` yet, so I didn't extend that.

**Comment by sffc** [comment]
> Added. IIRC we didn't add `Array.toLocaleString` to use `ListFormat` yet, so I didn't extend that.

Right; that's #422 :)

**Comment by zbraniecki** [comment]
We reached Stage 4 last week and this PR should be ready to be merged.

**Comment by zbraniecki** [comment]
@leobalter @gibson042 - ping?

File: spec/overview.html [hunk 1]
```diff
@@ -24,15 +24,15 @@ <h1>API Overview</h1>
...
     </p>
 
     <p>
       Applications can use the API in two ways:
 
       <ol>
         <li>
-          Directly, by using the constructors Intl.Collator, Intl.NumberFormat, Intl.DateTimeFormat, or Intl.PluralRules to construct an object, specifying a list of preferred languages and options to configure the behaviour of the resulting object. The object then provides a main function (compare, select, or format), which can be called repeatedly. It also provides a resolvedOptions function, which the application can use to find out the exact configuration of the object.
+          Directly, by using the constructors Intl.Collator, Intl.NumberFormat, Intl.DateTimeFormat, Intl.ListFormat or Intl.PluralRules to construct an object, specifying a list of preferred languages and options to configure the behaviour of the resulting object. The object then provides a main function (compare, select, or format), which can be called repeatedly. It also provides a resolvedOptions function, which the application can use to find out the exact configuration of the object.
```
Comment (leobalter) [inline spec/overview.html:35]:
```suggestion
          Directly, by using the constructors Intl.Collator, Intl.NumberFormat, Intl.DateTimeFormat, Intl.ListFormat, or Intl.PluralRules to construct an object, specifying a list of preferred languages and options to configure the behaviour of the resulting object. The object then provides a main function (compare, select, or format), which can be called repeatedly. It also provides a resolvedOptions function, which the application can use to find out the exact configuration of the object.
```

nit

File: spec/listformat.html [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+      <p>
+        The FormatListToParts abstract operation is called with arguments _listFormat_ (which must be an object initialized as a ListFormat) and _list_ (which must be a List of String values), and performs the following steps:
+      </p>
+
+      <emu-alg>
+        1. Let _parts_ be CreatePartsFromList(_listFormat_, _list_).
+        1. Let _result_ be ArrayCreate(0).
+        1. Let _n_ be 0.
+        1. For each _part_ in _parts_, do
+          1. Let _O_ be ObjectCreate(%ObjectPrototype%).
```
Comment (leobalter) [inline spec/listformat.html:123]:
```suggestion
          1. Let _O_ be ObjectCreate(%Object.prototype%).
```

Also, we need to update ObjectCreate in the whole spec text, as a follow up. It's now used as `OrdinaryObjectCreate`.

File: spec/listformat.html [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        Among the list types, `conjunction` stands for "and"-based lists (e.g., "A, B, and C"), `disjunction` stands for "or"-based lists (e.g., "A, B, or C"), and `unit` stands for lists of values with units (e.g., "5 pounds, 12 ounces").
+      </emu-note>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-properties-of-intl-listformat-prototype-object">
+    <h1>Properties of the Intl.ListFormat Prototype Object</h1>
+
+    <p>
+      The Intl.ListFormat prototype object is itself an ordinary object. <dfn>%ListFormatPrototype%</dfn> is not an Intl.ListFormat instance and does not have an [[InitializedListFormat]] internal slot or any of the other internal slots of Intl.ListFormat instance objects.
```
Comment (leobalter) [inline spec/listformat.html:273]:
```suggestion
      The Intl.ListFormat prototype object is itself an ordinary object. <dfn>%ListFormat.prototype%</dfn> is not an Intl.ListFormat instance and does not have an [[InitializedListFormat]] internal slot or any of the other internal slots of Intl.ListFormat instance objects.
```

File: spec/listformat.html [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+    <emu-clause id="sec-Intl.ListFormat">
+      <h1>Intl.ListFormat ( [ _locales_ [ , _options_ ] ] )</h1>
+
+      <p>
+        When the `Intl.ListFormat` function is called with optional arguments _locales_ and _options_, the following steps are taken:
+      </p>
+
+      <emu-alg>
+        1. If NewTarget is *undefined*, throw a *TypeError* exception.
+        1. Let _listFormat_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%ListFormatPrototype%"*, &laquo; [[InitializedListFormat]], [[Locale]], [[Type]], [[Style]], [[Templates]] &raquo;).
```
Comment (leobalter) [inline spec/listformat.html:175]:
```suggestion
        1. Let _listFormat_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%ListFormat.prototype%"*, &laquo; [[InitializedListFormat]], [[Locale]], [[Type]], [[Style]], [[Templates]] &raquo;).
```

File: spec/listformat.html [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+
+    <p>
+      The Intl.ListFormat constructor has the following properties:
+    </p>
+
+    <emu-clause id="sec-Intl.ListFormat.prototype">
+      <h1>Intl.ListFormat.prototype</h1>
+
+      <p>
+        The value of `Intl.ListFormat.prototype` is %ListFormatPrototype%.
```
Comment (leobalter) [inline spec/listformat.html:211]:
```suggestion
        The value of `Intl.ListFormat.prototype` is %ListFormat.prototype%.
```

File: spec/listformat.html [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        </table>
+      </emu-table>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-properties-of-intl-listformat-instances">
+    <h1>Properties of Intl.ListFormat Instances</h1>
+
+    <p>
+      Intl.ListFormat instances inherit properties from %ListFormatPrototype%.
```
Comment (leobalter) [inline spec/listformat.html:377]:
```suggestion
      Intl.ListFormat instances inherit properties from %ListFormat.prototype%.
```

File: spec/listformat.html [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        1. Let _lf_ be the *this* value.
+        1. If Type(_lf_) is not Object, throw a *TypeError* exception.
+        1. If _lf_ does not have an [[InitializedListFormat]] internal slot, throw a *TypeError* exception.
+        1. Let _stringList_ be ? StringListFromIterable(_list_).
+        1. Return FormatListToParts(_lf_, _stringList_).
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-Intl.ListFormat.prototype.resolvedoptions">
+      <h1>Intl.ListFormat.prototype.resolvedOptions ()</h1>
```
Comment (leobalter) [inline spec/listformat.html:328]:
```suggestion
      <h1>Intl.ListFormat.prototype.resolvedOptions ( )</h1>
```

File: spec/listformat.html [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+
+      <p>
+        This function provides access to the locale and options computed during initialization of the object.
+      </p>
+
+      <emu-alg>
+        1. Let _lf_ be the *this* value.
+        1. If Type(_lf_) is not Object, throw a *TypeError* exception.
+        1. If _lf_ does not have an [[InitializedListFormat]] internal slot, throw a *TypeError* exception.
+        1. Let _options_ be ! ObjectCreate(%ObjectPrototype%).
```
Comment (leobalter) [inline spec/listformat.html:338]:
```suggestion
        1. Let _options_ be ! ObjectCreate(%Object.prototype%).
```

**Review by leobalter — CHANGES_REQUESTED** [review]
We have some nit picks I can eventually address from my side.

cc @gibson042

**Review by leobalter — APPROVED** [review]

**Comment by leobalter** [comment]
All of the changes were editorial, minor fixes, there is no need to round this back and forth.

**Pull request merged**